### PR TITLE
ACS-4034: Assign a piece of content to a category

### PR DESCRIPTION
**Pull request by krdabrowski**
POST /nodes/{nodeId}/category-links
https://alfresco.atlassian.net/browse/ACS-4034
- added also one TAS test for update category covering repeated name case

[ACS-4034]: https://alfresco.atlassian.net/browse/ACS-4034?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: packaging/tests/tas-restapi/src/test/java/org/alfresco/rest/categories/LinkToCategoriesTests.java [hunk 1]
```diff
@@ -0,0 +1,276 @@
...
+        fileNode.assertThat().field(PROPERTIES_FIELD).contains(thirdCategory.getId());
+    }
+
+    /**
+     * Try to link content to category as an improper user and expect 403 (Forbidden)
+     */
+    @Test(groups = { TestGroup.REST_API})
+    public void testLinkContentToCategory_usingUserWithoutAccessRightsAndExpect403()
+    {
+        STEP("Try to link content to a category using user without read permission and expect 403");
```
Comment (tpage-alfresco) [inline packaging/tests/tas-restapi/src/test/java/org/alfresco/rest/categories/LinkToCategoriesTests.java:183]: I think we should try a user without change permission instead (and expect 403).

File: packaging/tests/tas-restapi/src/test/java/org/alfresco/rest/categories/LinkToCategoriesTests.java [hunk 2]
```diff
@@ -0,0 +1,276 @@
...
+        STEP("Try to call link content API with empty category ID and expect 400");
+        final String nonExistingCategoryId = StringUtils.EMPTY;
+        final RestCategoryLinkBodyModel categoryLink = createCategoryLinkWithId(nonExistingCategoryId);
+        restClient.authenticateUser(user).withCoreAPI().usingNode(file).linkToCategory(categoryLink);
+
+        restClient.assertStatusCodeIs(BAD_REQUEST);
+    }
+
+    /**
+     * Try to link non-content node to category and expect 400 (Bad Request)
```
Comment (tpage-alfresco) [inline packaging/tests/tas-restapi/src/test/java/org/alfresco/rest/categories/LinkToCategoriesTests.java:232]: I'm surprised at this. I thought that Share allowed folders to be linked to categories.

File: packaging/tests/tas-restapi/src/test/java/org/alfresco/rest/categories/UpdateCategoriesTests.java [hunk 3]
```diff
@@ -93,6 +94,27 @@ public void testUpdateSubcategory_asAdmin()
         updatedCategory.assertThat().field(FIELD_NAME).is(categoryNewName);
     }
 
+    /**
+     * Try to update a category with a name, which is already present within the parent category
+     */
+    @Test(groups = { TestGroup.REST_API})
+    public void testUpdateCategory_usingRecurringName()
```
Comment (tpage-alfresco) [inline packaging/tests/tas-restapi/src/test/java/org/alfresco/rest/categories/UpdateCategoriesTests.java:101]: This looks like a good extra test, but it seems unrelated to the change. I don't have an objection - more tests is good as you think of them, I just wanted to check that you intended to include this.

File: remote-api/src/main/java/org/alfresco/rest/api/impl/CategoriesImpl.java [hunk 4]
```diff
@@ -236,15 +292,68 @@ private NodeRef changeCategoryName(final NodeRef categoryNodeRef, final String n
...
+    {
+        if (currentCategories == null)
+        {
+            return newCategories;
+        }
+
+        final Collection<NodeRef> allCategories = DefaultTypeConverter.INSTANCE.getCollection(NodeRef.class, currentCategories);
+        newCategories.stream()
+            .filter(category -> !allCategories.contains(category))
+            .forEach(allCategories::add);
```
Comment (tpage-alfresco) [inline remote-api/src/main/java/org/alfresco/rest/api/impl/CategoriesImpl.java:334]: Do we know that `allCategories` is modifiable? Instead we could use the union of two sets to determine what the new value should contain.

File: remote-api/src/main/java/org/alfresco/rest/api/impl/CategoriesImpl.java [hunk 5]
```diff
@@ -236,15 +292,68 @@ private NodeRef changeCategoryName(final NodeRef categoryNodeRef, final String n
...
+        return allCategories;
+    }
+
+    /**
+     * Add to or update node's property cm:categories containing linked category references.
+     *
+     * @param nodeRef Node reference.
+     * @param categoryNodeRefs Category node references.
+     */
+    private void linkNodeToCategories(final NodeRef nodeRef, final Collection<NodeRef> categoryNodeRefs)
```
Comment (tpage-alfresco) [inline remote-api/src/main/java/org/alfresco/rest/api/impl/CategoriesImpl.java:345]: This looks fine to me, but can't we reuse some existing service code for this? (Maybe in the CategoryService?)

**Review by mpichura — COMMENTED** [review]
I think we are not having parity with some Share functionality (applying categories to folders, possibly more restrictive check on node permissions).

File: packaging/tests/tas-restapi/src/test/java/org/alfresco/rest/categories/LinkToCategoriesTests.java (hunk 2 above)
Comment (tpage-alfresco) [reply packaging/tests/tas-restapi/src/test/java/org/alfresco/rest/categories/LinkToCategoriesTests.java:232]: Yes - it's probably a good idea to look to see if there's any restriction in the type of node in the existing v0 webscript.

File: remote-api/src/main/java/org/alfresco/rest/api/impl/CategoriesImpl.java (hunk 5 above)
Comment (tpage-alfresco) [reply remote-api/src/main/java/org/alfresco/rest/api/impl/CategoriesImpl.java:345]: Ok - that sounds like a good idea, but if it's a more general change then we should get this PR merged first.

File: packaging/tests/tas-restapi/src/test/java/org/alfresco/rest/categories/LinkToCategoriesTests.java (hunk 1 above)
Comment (krdabrowski) [reply packaging/tests/tas-restapi/src/test/java/org/alfresco/rest/categories/LinkToCategoriesTests.java:183]: I don't know why but I assumed that read permission will be enough to create a link. Now after rethinking you are right - I will fix it.

File: packaging/tests/tas-restapi/src/test/java/org/alfresco/rest/categories/LinkToCategoriesTests.java (hunk 2 above)
Comment (krdabrowski) [reply packaging/tests/tas-restapi/src/test/java/org/alfresco/rest/categories/LinkToCategoriesTests.java:232]: I've just verified that and you are true. Share in did allows folder linking. I will fix that.

File: remote-api/src/main/java/org/alfresco/rest/api/impl/CategoriesImpl.java (hunk 4 above)
Comment (krdabrowski) [reply remote-api/src/main/java/org/alfresco/rest/api/impl/CategoriesImpl.java:334]: The `.getCollection()` always creates a new `ArrayList<T>` so yes - it's modifiable. In the other hand it's nice suggestion to use sets union - I will change my solution.

File: remote-api/src/main/java/org/alfresco/rest/api/impl/CategoriesImpl.java (hunk 5 above)
Comment (krdabrowski) [reply remote-api/src/main/java/org/alfresco/rest/api/impl/CategoriesImpl.java:345]: As Maciej already mentioned, unfortunately there doesn't exist any service code which could be reused. Similar solution is used in `LinkCategoryActionExecuter` and we could update it also there if we would introduce some new "CategoryService".

File: packaging/tests/tas-restapi/src/test/java/org/alfresco/rest/categories/LinkToCategoriesTests.java (hunk 2 above)
Comment (krdabrowski) [reply packaging/tests/tas-restapi/src/test/java/org/alfresco/rest/categories/LinkToCategoriesTests.java:232]: Regarding other node types - v0 API allows to link different kinds of nodes to categories. I don't think there's any check for that. I've asked Stefan if we should support only Content and Folder nodes.

Comment (tpage-alfresco) [reply packaging/tests/tas-restapi/src/test/java/org/alfresco/rest/categories/LinkToCategoriesTests.java:232]: It's a good idea to check. For now I would take the check out completely so that this can be merged. Then we can have a new ticket for the validation once we know what the requirements are.

File: remote-api/src/main/java/org/alfresco/rest/api/impl/CategoriesImpl.java [hunk 6]
```diff
@@ -141,6 +152,35 @@ public void deleteCategoryById(String id, Parameters parameters)
...
+    public List<Category> linkContentNodeToCategories(final String nodeId, final List<Category> categoryLinks)
+    {
+        if (CollectionUtils.isEmpty(categoryLinks))
+        {
+            throw new InvalidArgumentException(NOT_A_VALID_CATEGORY);
+        }
+
+        final NodeRef contentNodeRef = nodes.validateNode(nodeId);
+        verifyReadPermission(contentNodeRef);
+        verifyNodeType(contentNodeRef, ContentModel.TYPE_CONTENT);
```
[the remark this answers is not in the document]
Comment (krdabrowski) [reply remote-api/src/main/java/org/alfresco/rest/api/impl/CategoriesImpl.java]: Fixed - node type verification removed.

**Comment by krdabrowski** [comment]
After talk with @jamalkm added node type verification. It should look the same way like for tags, see: https://github.com/Alfresco/alfresco-community-repo/blob/master/remote-api/src/main/java/org/alfresco/rest/api/impl/TagsImpl.java#L86

File: packaging/tests/tas-restapi/src/test/java/org/alfresco/rest/categories/LinkToCategoriesTests.java [hunk 7]
```diff
@@ -0,0 +1,350 @@
...
+    public void testLinkContentToCategory_usingTagInsteadOfContentAndExpect405()
+    {
+        STEP("Try to link a tag to category and expect 405");
+        final RestCategoryLinkBodyModel categoryLink = createCategoryLinkWithId(category.getId());
+        final RestTagModel tag = restClient.authenticateUser(user).withCoreAPI().usingNode(file).addTag("someTag");
+        final RepoTestModel tagNode = new RepoTestModel() {};
+        tagNode.setNodeRef(tag.getId());
+        restClient.authenticateUser(dataUser.getAdminUser()).withCoreAPI().usingNode(tagNode).linkToCategory(categoryLink);
+
+        restClient.assertStatusCodeIs(METHOD_NOT_ALLOWED);
```
Comment (krdabrowski) [inline packaging/tests/tas-restapi/src/test/java/org/alfresco/rest/categories/LinkToCategoriesTests.java:301]: `METHOT_NOT_ALLOWED` - to maintain consistency with tags API.

File: packaging/tests/tas-restapi/src/test/java/org/alfresco/rest/categories/LinkToCategoriesTests.java [hunk 8]
```diff
@@ -0,0 +1,350 @@
...
+        category = prepareCategoryUnderRoot();
+    }
+
+    /**
+     * Link content to category and verify if this category is present in node's properties
+     */
+    @Test(groups = { TestGroup.REST_API})
+    public void testLinkContentToCategory()
+    {
+        STEP("Check if file is not linked to any category");
```
Comment (tpage-alfresco) [inline packaging/tests/tas-restapi/src/test/java/org/alfresco/rest/categories/LinkToCategoriesTests.java:92]: It would be better to use a new file for this test, otherwise other tests may cause this to fail.

Comment (tpage-alfresco) [reply packaging/tests/tas-restapi/src/test/java/org/alfresco/rest/categories/LinkToCategoriesTests.java:92]: Ah - I just noticed it's created in `BeforeMethod` - ignore this then.

File: packaging/tests/tas-restapi/src/test/java/org/alfresco/rest/categories/LinkToCategoriesTests.java [hunk 9]
```diff
@@ -0,0 +1,350 @@
...
+        STEP("Use admin to create a private site");
+        final SiteModel privateSite = dataSite.usingAdmin().createPrivateRandomSite();
+
+        STEP("Add the user to the site, let him create a folder and then evict him from the site again");
+        dataUser.addUserToSite(user, privateSite, SiteManager);
+        final FolderModel privateFolder = dataContent.usingUser(user).usingSite(privateSite).createFolder();
+        final FileModel privateFile = dataContent.usingUser(user).usingResource(privateFolder).createContent(CMISUtil.DocumentType.TEXT_PLAIN);
+        dataUser.removeUserFromSite(user, privateSite);
+
+        STEP("Try to link content to a category as owner and expect 403");
```
Comment (tpage-alfresco) [inline packaging/tests/tas-restapi/src/test/java/org/alfresco/rest/categories/LinkToCategoriesTests.java:228]: I don't understand this. The owner should be able to link the content to a category.

File: packaging/tests/tas-restapi/src/test/java/org/alfresco/rest/categories/LinkToCategoriesTests.java (hunk 7 above)
Comment (tpage-alfresco) [reply packaging/tests/tas-restapi/src/test/java/org/alfresco/rest/categories/LinkToCategoriesTests.java:301]: This is fine, but is it documented in the api-explorer?

File: remote-api/src/main/resources/alfresco/public-rest-context.xml [hunk 10]
```diff
@@ -834,6 +834,8 @@
         <constructor-arg name="categoryService" ref="CategoryService"/>
         <constructor-arg name="nodes" ref="nodes"/>
         <constructor-arg name="nodeService" ref="NodeService"/>
+        <constructor-arg name="permissionService" ref="PermissionService"/>
+        <constructor-arg name="typeConstraint" ref="nodeTypeConstraint"/>
```
Comment (tpage-alfresco) [inline remote-api/src/main/resources/alfresco/public-rest-context.xml:838]: This is a good discovery :+1:

File: packaging/tests/tas-restapi/src/test/java/org/alfresco/rest/categories/LinkToCategoriesTests.java (hunk 7 above)
Comment (mpichura) [reply packaging/tests/tas-restapi/src/test/java/org/alfresco/rest/categories/LinkToCategoriesTests.java:301]: I see the point of being consistent with Tags API but I don't think this status code is meant for that. AFAIR, it should be returned when http method is not handled/accepted by the server. 
TBH, I don't think we should repeat mistake we made with tags and we should either return `400 Bad request` or `422 Unprocessable entity`

Comment (mpichura) [reply packaging/tests/tas-restapi/src/test/java/org/alfresco/rest/categories/LinkToCategoriesTests.java:301]: As discussed - I believe we can move on with this code and log a ticket to get this on and (possibly tags) fixed to return a more suitable error code.

**Review by mpichura — APPROVED** [review]
Apart from the error code 405 part which we hopefully will handle soon, LGTM.

File: packaging/tests/tas-restapi/src/test/java/org/alfresco/rest/categories/LinkToCategoriesTests.java (hunk 7 above)
Comment (krdabrowski) [reply packaging/tests/tas-restapi/src/test/java/org/alfresco/rest/categories/LinkToCategoriesTests.java:301]: Ticked created: https://alfresco.atlassian.net/browse/ACS-4362

**Review by tpage-alfresco — APPROVED** [review]
LGTM. Sorry - I thought I'd already +1ed this.
